### PR TITLE
Better handling for parseExceptions for Batch Ingestion

### PR DESCRIPTION
**Pull request by nishantmonu51**
Issue reported here - 
https://groups.google.com/d/msgid/druid-development/220cbbfe-a469-4d4c-b473-14e650dafba5%40googlegroups.com

### CODE REVIEW

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java [hunk 1]
```diff
@@ -64,26 +65,22 @@ protected void map(
...
-
+      final InputRow inputRow = parseInputRow(value, parser);
       if (!granularitySpec.bucketIntervals().isPresent()
           || granularitySpec.bucketInterval(new DateTime(inputRow.getTimestampFromEpoch()))
                             .isPresent()) {
         innerMap(inputRow, value, context);
       }
     }
+    catch (ParseException e) {
+      if (config.isIgnoreInvalidRows()) {
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java]: can we flip this around for readability's sake and put the `isIgnoreInvalidRows` outside the `if`?

**Comment by gianm** [comment]
It'd be more consistent with indexing to fix this by using a similar construction to IncrementalIndex rather than by catching ParseExceptions on "innerMap",

```java
        try {
          agg.aggregate();
        }
        catch (ParseException e) {
          // "aggregate" can throw ParseExceptions if a selector expects something but gets something else.
          if (reportParseExceptions) {
            throw new ParseException(e, "Encountered parse error for aggregator[%s]", agg.getName());
          } else {
            log.debug(e, "Encountered parse error, skipping aggregator[%s].", agg.getName());
          }
        }
```

**Comment by gianm** [comment]
that behavior would mean if a single metric is bad, but the rest of the row is ok, we would include all but the bad metric, rather than throwing away the entire row.

**Comment by gianm** [comment]
using that construction would also mean that the adjustment of the try/catch in HadoopDruidIndexerMapper isn't necessary (as ParseExceptions will not be thrown if reportParseExceptions is false)

**Comment by nishantmonu51** [comment]
@gianm @drcrallen - changed it to be consistent with RT. please check again.

**Comment by gianm** [comment]
👍

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java [hunk 2]
```diff
@@ -68,20 +71,20 @@ protected void map(
       try {
         inputRow = parseInputRow(value, parser);
       }
-      catch (Exception e) {
-        if (config.isIgnoreInvalidRows()) {
+      catch (ParseException e) {
+        if (reportParseExceptions) {
+          throw e;
+        } else {
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java]: You can get rid of the nesting in an else here

File: indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java [hunk 3]
```diff
@@ -91,7 +92,17 @@ public InputRow get()
...
         );
-        agg.aggregate();
+        try {
+          agg.aggregate();
+        }
+        catch (ParseException e) {
+          // "aggregate" can throw ParseExceptions if a selector expects something but gets something else.
+          if (reportParseExceptions) {
+            throw new ParseException(e, "Encountered parse error for aggregator[%s]", agg.getName());
+          } else {
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java]: same here with being able to get rid of nesting of else

File: indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java [hunk 4]
```diff
@@ -91,7 +92,17 @@ public InputRow get()
                 true
             )
         );
-        agg.aggregate();
+        try {
+          agg.aggregate();
+        }
+        catch (ParseException e) {
+          // "aggregate" can throw ParseExceptions if a selector expects something but gets something else.
+          if (reportParseExceptions) {
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java]: isn't this backwards? should be `reportParseExceptions`?

if so needs unit test

Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java]: ah nevermind. I had it backwards in my head

File: indexing-hadoop/src/test/java/io/druid/indexer/InputRowSerdeTest.java [hunk 5]
```diff
@@ -88,5 +90,29 @@ public void testSerde()
...
+        new DoubleSumAggregatorFactory("agg_non_existing", "agg_non_existing_in"),
+        new DoubleSumAggregatorFactory("m1out", "m1"),
+        new LongSumAggregatorFactory("m2out", "m2"),
+        new HyperUniquesAggregatorFactory("m3out", "m3"),
+        new LongSumAggregatorFactory("unparseable", "m3") // Unparseable from String to Long
+    };
+    try {
+      InputRowSerde.toBytes(in, aggregatorFactories, true);
+      Assert.fail("Should have thrown parseException");
+    } catch (ParseException e){
```
Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/InputRowSerdeTest.java]: Better to use the `org.junit.rules.ExpectedException` `@Rule` or use `@Test(expected = ParseException.class)`

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java (hunk 2 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java]: done.

File: indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java (hunk 3 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java]: done.

**Comment by drcrallen** [comment]
👍

**Comment by drcrallen** [comment]
miss-click sorry about that

**Pull request merged**